lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/AvroFileReader.java

```diff
@@ -0,0 +1,176 @@
+import java.util.function.Function;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+
+public class AvroFileReader
+        implements Closeable
+{
+    private final TrinoDataInputStream input;
+    private final AvroPageDataReader dataReader;
+    private final DataFileReader<Optional<Page>> fileReader;
+    private Page nextPage;
+    private final OptionalLong end;
+
+    public AvroFileReader(
+            TrinoInputFile inputFile,
+            Schema schema,
+            AvroTypeManager avroTypeManager)
+            throws IOException, AvroTypeException
+    {
+        this(inputFile, schema, avroTypeManager, 0, OptionalLong.empty());
+    }
+
+    public AvroFileReader(
+            TrinoInputFile inputFile,
+            Schema schema,
+            AvroTypeManager avroTypeManager,
+            long offset,
+            OptionalLong length)
+            throws IOException, AvroTypeException
+    {
+        requireNonNull(inputFile, "inputFile is null");
+        requireNonNull(schema, "schema is null");
+        requireNonNull(avroTypeManager, "avroTypeManager is null");
+        long fileSize = inputFile.length();
+
+        verify(offset >= 0, "offset is negative");
+        verify(offset < inputFile.length(), "offset is greater than data size");
+        length.ifPresent(lengthLong -> verify(lengthLong >= 1, "length must be at least 1"));
+        end = length.stream().map(l -> l + offset).findFirst();
+        end.ifPresent(endLong -> verify(endLong <= fileSize, "offset plus length is greater than data size"));
+        input = new TrinoDataInputStream(inputFile.newStream());
+        dataReader = new AvroPageDataReader(schema, avroTypeManager);
+        try {
+            fileReader = new DataFileReader<>(new TrinoDataInputStreamAsAvroSeekableInput(input, fileSize), dataReader);
+            fileReader.sync(offset);
+        }
+        catch (AvroPageDataReader.UncheckedAvroTypeException runtimeWrapper) {
+            // Avro Dataum Reader interface can't throw checked exceptions when initialized by the file reader,
+            // so the exception is wrapped in a runtime exception that must be unwrapped
+            throw runtimeWrapper.getAvroTypeException();
+        }
+        avroTypeManager.configure(fileReader.getMetaKeys().stream().collect(toImmutableMap(Function.identity(), fileReader::getMeta)));
+    }
+
+    public long getCompletedBytes()
+    {
+        return input.getReadBytes();
+    }
+
+    public long getReadTimeNanos()
+    {
+        return input.getReadTimeNanos();
+    }
+
+    public boolean hasNext()
+            throws IOException
+    {
+        loadNextPageIfNecessary();
+        return Objects.nonNull(nextPage);
+    }
+
+    public Page next()
+            throws IOException
+    {
+        loadNextPageIfNecessary();
+        Page result = nextPage;
+        nextPage = null;
```

Comment:
Regarding your comment here and below:
should we store a local `Optional<Page>` and throw an exception if its empty after a `loadNextPageIfNecessary` call? That way we could always just call `Optional::get` for the resulting page without any `orElse(null)`
Or just return an Optional from `next` that is full only in cases where `hasNext` is true? it will require pushing the `orElse(null)` into the PageSource

Reply:
@dain's comment is about the interface of this class. It should either

* have `hasNext()` with `next()` throwing if there is no next page
* only have `next()` which returns either a nullable `Page` or `Optional<Page>`

Returning `Optional<Page>` seems simplest as then `loadNextPageIfNecessary()` can probably be inlined into this method.

Reply:
The weirdness being the underlying Avro data reader doesn't follow either of those patterns. It's a series of optionals where most are empty and then every once in a while there is a page.